### ISSUE TITLE
AwaitTaskCorrect

### ISSUE THREAD

**Issue by edgarjrg** [fsharp/fslang-suggestions#840]
# Title of Suggestion

I propose including [AwaitTaskCorrect](http://www.fssnip.net/7Rc/title/AsyncAwaitTaskCorrect) (maybe with another name)

```F#
open System
open System.Threading.Tasks

type Async with
    static member AwaitTaskCorrect(task : Task) : Async<unit> =
        Async.FromContinuations(fun (sc,ec,cc) ->
            task.ContinueWith(fun (task:Task) ->
                if task.IsFaulted then
                    let e = task.Exception
                    if e.InnerExceptions.Count = 1 then ec e.InnerExceptions.[0]
                    else ec e
                elif task.IsCanceled then
                    ec(TaskCanceledException())
                else
                    sc ())
            |> ignore)

    static member AwaitTaskCorrect(task : Task<'T>) : Async<'T> =
        Async.FromContinuations(fun (sc,ec,cc) ->
            task.ContinueWith(fun (task:Task<'T>) ->
                if task.IsFaulted then
                    let e = task.Exception
                    if e.InnerExceptions.Count = 1 then ec e.InnerExceptions.[0]
                    else ec e
                elif task.IsCanceled then
                    ec(TaskCanceledException())
                else
                    sc task.Result)
            |> ignore)


// examples

let mkFailingTask exn = Task.Factory.StartNew<_>(fun () -> raise exn)

let test1 taskAwaiter =
    async {
        try
            return! taskAwaiter (mkFailingTask (ArgumentException()))
        with
        | :? ArgumentException -> return true
        | :? AggregateException -> return false
    } |> Async.RunSynchronously


test1 Async.AwaitTask // false
test1 Async.AwaitTaskCorrect // true


let test2 taskAwaiter =
    async {
        try
            return! taskAwaiter (mkFailingTask (AggregateException("kaboom!")))
        with
        | :? AggregateException as e when e.Message = "kaboom!" -> return true
        | :? AggregateException -> return false
    } |> Async.RunSynchronously

test2 Async.AwaitTask // false
test2 Async.AwaitTaskCorrect // true
```

The existing way of approaching this problem in F# is copying the snippet in your project, ref :  jet/equinox#198

## Pros and Cons 

The advantages of making this adjustment to F#:
- Provides a better implementation of Async.AwaitTask where the actual exception of a failing task is passed to the async mechanism.
- Works as C# would handle taks

The disadvantages of making this adjustment to F# are:
- Might be confusing having 2 AwaitTask 
- If AwaitTask is "fixed" then is a breaking change

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): No idea

Related suggestions: at this moment I don't know of any other related suggestions might edit in the future

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [X] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [X] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [X] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [X] This is not a breaking change to the F# language design
* [X] I or my company would be willing to help implement and/or test this

**Comment by FraserWaters-GR** [comment]
If breaking AwaitTask it would be good to fix up another of it's oddities at the same time, that it can't be cancelled: https://github.com/dotnet/fsharp/pull/7357

**Comment by matthid** [comment]
This is related to https://github.com/dotnet/fsharp/pull/3257 
Indeed, I discussed this quirk there as well.
I'm skeptical about changing the exception type here... 
I'm also a bit skeptical about calling it `Correct`
Maybe we can add an Overload and a Method parameter (`unwrapException`)?

**Comment by edgarjrg** [comment]
@matthid I’ll agree that the “Correct” part has to go, I was just honoring the Original Poster of the snippet. 

The overload with a param that triggers the new behavior seems fine to me.

I checked that PR real quick and I’m not sure if it is exactly this, but really related should have been merged 😱 

How do you deal with this? Is there any library that compensates this design right now?

**Comment by matthid** [comment]
> How do you deal with this? Is there any library that compensates this design right now?

I usually have some extension-methods to check for inner exceptions and/or unwrap the exception if needed

**Comment by dsyme** [comment]
I agree this should be solved.  We need a better name - adding an optional argument seems reasonable though is a binary breaking change (though I believe it can be arranged that it's not a source breaking change on re-compilation).

I will mark this as approved in principle - would be great to have it as a community-led contribution

For the `IsFaulted` path should the code check that AggregateException is being thrown?

**Comment by abelbraaksma** [comment]
Related: https://github.com/fsprojects/FSharp.Control.TaskSeq/issues/128.

**Comment by dsyme** [comment]
We need a better name.  Perhaps if we had an overloaded Async.Await(task) overloaded with Async, Task, ValueTask and Task-like in FSharp.Core it would be best

**Comment by abelbraaksma** [comment]
Note that `Async.Await(ValueTask)` will be NS2.1 only.

Is it wise to put all these overloads in `Async`? It may be hard to discover, people would probably expect task-like operations to be in `Task.xxx`, and same for `ValueTask`.

**Comment by T-Gro** [comment]
We could add them symmetrically, similarly to how it is done for collection conversions

Async.Await(ValueTask)
ValueTask.toAsync

**Comment by TheAngryByrd** [comment]
I decided to add them in [IcedTasks](https://github.com/TheAngryByrd/IcedTasks/pull/24) under `AsyncEx` if anyone wants to give it a try.

**Comment by gusty** [comment]
What about other Async functions doing surprising wrapping of aggregate exceptions?

For instance, I think we need something like:

```
type Async with
    static member StartImmediateAsTaskCorrect (computation: Async<'T>, ?cancellationToken) : Task<'T> =
        let cancellationToken = defaultArg cancellationToken (new CancellationTokenSource ()).Token
        let ts = TaskCompletionSource<'T> ()
        Async.StartWithContinuations (
            computation,
            ts.SetResult,
            (function
                | :? AggregateException as agg -> ts.SetException agg.InnerExceptions
                | exn -> ts.SetException exn),
            (fun _ -> ts.SetCanceled ()),
            cancellationToken)
        ts.Task
```

Maybe I should create a separate issue.

**Comment by bartelink** [comment]
Think that (similar warts/concerns re StartImmediateAsTask/StartAsTask etc)'s best covered under a separate issue (and revisiting `Async.DefaultCancellationToken` stuff can be talked about there too...). I feel this thread is plenty busy just focusing on an `AwaitTask` with corrected wait semantics vs bringing in the starting of work aspects into play.

**Comment by T-Gro** [comment]
This does not necessarily need a full RFC, but it needs a full API diff proposal - what's added where and elaborate on binary and source compat for code expecting an AggregateException.

If we wanted new code to stup using existing members, but keep it in for binary compat, support for OverloadResolutionPriorityAttribute could be the mechanism to apply.

I tried to come up with a different naming or an `open`-based solution, but none of it feels right.

**Comment by bartelink** [comment]
Re naming:

For me, while `Async.AwaitTask` is in all the examples and prod code, correcting the semantics, or having to do something ugly to opt into sensible ones is simply bad news (despite the fact that the semantics being broken initially came as news to the maintainers of TaskSeq and FsToolkit when I flagged it). So, I think `Await` is a reasonable name; that said having a plan as to how people become aware of it existing and being worth migrating to is definitely part of the job. From a tooling perspective having a different name is also going to help with LLMs in the mix.

Creative solutions like shadowing `AsyncBuilder` are already in the wild, but for me trying to be too clever in `FSharp.Core` is a recipe for a mess. For better or worse, there are lots of closed source codebases that have `Async.AwaitTask` sprinkled throughout, with and without compensating unwrapping of `AggregateException`s and/or correct behavior in exception paths. [Hyrum's law](https://www.hyrumslaw.com) answers the question of whether it's worth making a breaking change to semantics in the lib. Reasonable people will differ on whether tooling, shadowing builders, or explicit find and replace are the way to move codebases onto the new semantics at an appropriate time with people being aware of the nature of the change. 

----

Re proposed signatures/diffs (assuming there's agreement to mint a new name), [Don's suggestions above](https://github.com/fsharp/fslang-suggestions/issues/840#issuecomment-1376025586) seem to be a good starting point:

The key ones we should focus on under this issue IMO are:

1. `Async.Await(Task)` -- as per snippet in OP
2. `Async.Await(Task<'T>)` -- as per snippet in OP

> `Async.Await(Async)`

@dsyme can you expand on what the semantics of that would be? (i.e. does it start something running and if so how/when)

> `Async.Await(ValueTask)`
> `Async.Await(task-like)`

Other overloads where justified by perf or reducing allocations are fine, but I personally would not fixate on them and/or let discussing them delay the resolution of the OP as:

1. The actual core need here is to remove a long standing pitfall of surprising exception wrapping from less trafficked execution paths that has occupied way too much airtime in real world usage of the language
2. `ValueTask.AsTask` is not quite the level of perf sin given the current `async { }` perf/allocation characteristics
3. there are lots of surrounding issues esp wrt providing a complete story for ValueTask etc, and providing appropriate overloads on `Async.Await` can live with those.

**Comment by bartelink** [comment]
FYI the impl of this is now in trunk via https://github.com/dotnet/fsharp/pull/19785 - probably makes sense to close it to take it off the list (as we've done with #1466)
